### PR TITLE
dts: flash_map: Automatically generate of the list of flash area

### PR DESCRIPTION
**Pull request by galak**
This Patch add functionality for automatic generation of the flash map
using DTS description. Automatic generation allows to replace
C-hardcoded flash_map.

We generate a set of defines based on the index of a partiion:
	#define DT_FLASH_AREA_<IDX>_OFFSET	0
	#define DT_FLASH_AREA_<IDX>_SIZE	131072
	#define DT_FLASH_AREA_<IDX>_DEV		"FLASH_CTRL"
	#define DT_FLASH_AREA_<IDX>_LABEL	MCUBOOT

Additionally we also define:
	#define DT_FLASH_AREA_NUM		4

and:
	#define DT_FLASH_AREA_<PARTNAME>_ID	0

Signed-off-by: Findlay Feng <i@fengch.me>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Comment by galak** [comment]
@findlayfeng @nvlsianpu this is take the base 3 commits from @findlayfeng PR #8837 and reworks the basic dt extraction support.

I've built tested this but haven't run it on HW to make sure flash map is working properly.

File: include/flash_map.h [hunk 1]
```diff
@@ -42,6 +42,7 @@ struct flash_area {
 	u16_t pad16;
 	off_t fa_off;
 	size_t fa_size;
+	const char *fa_name;
```
Comment (nvlsianpu) [inline include/flash_map.h]: I believe it fa_device_id is not needed anymore

Comment (nvlsianpu) [reply include/flash_map.h]: fa_name -> fa_dev_name

**Review by nvlsianpu — COMMENTED** [review]
will test this

**Comment by galak** [comment]
> @galak, would it not be better to use fa.dev_name instead of fa.name ?

Yeah, that makes more sense.

File: include/flash_map.h (hunk 1 above)
Comment (galak) [reply include/flash_map.h]: > I believe it fa_device_id is not needed anymore

Its not, was keeping it around until we decide on deprecation policy vs removal for out of tree flash_maps.

**Comment by nvlsianpu** [comment]
With this patch flash area ID are not predefined - so need to use DT defined labels instead of explicite area numbers whenever call flash_area_open. FCB test needs corrections.

**Comment by galak** [comment]
> With this patch flash area ID are not predefined - so need to use DT defined labels instead of explicite area numbers whenever call flash_area_open. FCB test needs corrections.

Pushed changes to use DT_FLASH_AREA_* for in-tree users except `tests/subsys/settings/fcb/src/settings_test_fcb.c` & `subsys/settings/src/settings_init.c` as they use `CONFIG_SETTINGS_FCB_FLASH_AREA`.

**Review by nvlsianpu — APPROVED** [review]
looks good, works fine.

**Pull request merged**